File: llama_mobile-android-java-SDK/src/androidTest/java/com/llamamobile/LlamaMobileInstrumentedTests.java
package com.llamamobile;

import android.content.Context;
import android.content.res.AssetManager;
import android.os.Build;
import android.os.Environment;
import android.util.Log;

import androidx.test.platform.app.InstrumentationRegistry;
import androidx.test.ext.junit.runners.AndroidJUnit4;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import static org.junit.Assert.*;

import java.io.File;
import java.io.InputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.Arrays;

/**
 * Instrumented tests for LlamaMobile Java SDK
 * These tests run on an Android device or emulator.
 */
@RunWith(AndroidJUnit4.class)
public class LlamaMobileInstrumentedTests {

    private static final String TAG = "LlamaMobileTests";
    private static final String TEST_ASSET_DIR = "grammars";
    private static final String TEST_GRAMMAR_FILE = "json.gbnf";
    private String TEST_MODEL_DIR;
    private Context appContext;
    
    @Before
    public void setUp() {
        appContext = InstrumentationRegistry.getInstrumentation().getTargetContext();
        
        // Use internal device storage instead of sdcard
        File externalFilesDir = appContext.getExternalFilesDir(null);
        TEST_MODEL_DIR = new File(externalFilesDir, "llama_mobile/models").getAbsolutePath();
        
        // Create model directory structure if it doesn't exist
        File modelDir = new File(TEST_MODEL_DIR);
        if (!modelDir.exists()) {
            modelDir.mkdirs();
        }
    }

    @Test
    public void testAssetLoading() {
        AssetManager assetManager = appContext.getAssets();

        try {
            // List all grammar files
            String[] grammarFiles = assetManager.list(TEST_ASSET_DIR);
            assertNotNull("Grammar directory should exist", grammarFiles);
            Log.d(TAG, "Found grammar files: " + Arrays.toString(grammarFiles));
            
            // Check if specific grammar file exists
            boolean hasJsonGrammar = false;
            for (String file : grammarFiles) {
                if (file.equals(TEST_GRAMMAR_FILE)) {
                    hasJsonGrammar = true;
                    break;
                }
            }
            assertTrue("JSON grammar file should exist", hasJsonGrammar);
            
        } catch (IOException e) {
            Log.e(TAG, "Error accessing assets: " + e.getMessage());
            fail("Asset loading should succeed");
        }
    }

    @Test
    public void testFileDirectoryAccess() {
        // Test if we can create a directory for models
        File modelDir = new File(TEST_MODEL_DIR);
        boolean dirCreated = modelDir.mkdirs() || modelDir.exists();
        
        // This might fail if we don't have write permission, but we can still check if directory exists
        Log.d(TAG, "Model directory exists: " + modelDir.exists());
        Log.d(TAG, "Can write to model directory: " + modelDir.canWrite());
        
        // Just check if external storage is available
        String state = Environment.getExternalStorageState();
        assertTrue("External storage should be available", 
                Environment.MEDIA_MOUNTED.equals(state) || Environment.MEDIA_MOUNTED_READ_ONLY.equals(state));
    }

    @Test
    public void testNativeLibraryLoading() {
        try {
            // The library should be loaded automatically via static block
            Log.d(TAG, "Native library loaded successfully");
            
            // Test that we can access native methods
            long invalidContext = -1;
            assertFalse("Invalid context should not be valid", LlamaMobile.isContextValid(invalidContext));
            
        } catch (UnsatisfiedLinkError e) {
            Log.e(TAG, "Native library loading failed: " + e.getMessage());
            // This might fail on some test environments, so we don't fail the test
            Log.w(TAG, "Skipping native library test - this is expected on some environments");
        } catch (Exception e) {
            Log.e(TAG, "Unexpected error: " + e.getMessage());
            fail("Native library test should not throw exceptions");
        }
    }

    @Test
    public void testContextSafety() {
        // Test with invalid context
        long invalidContext = -1;
        assertFalse("Invalid context should not be valid", LlamaMobile.isContextValid(invalidContext));
        assertFalse("Should fail to release invalid context", LlamaMobile.releaseContext(invalidContext));
        
        // These should be safe to call with invalid context
        LlamaMobile.generateEmbedding(invalidContext, "test");
        LlamaMobile.generateCompletion(invalidContext, "test");
        
        Log.d(TAG, "Context safety tests passed");
    }

    @Test
    public void testDeviceCompatibility() {
        Context appContext = InstrumentationRegistry.getInstrumentation().getTargetContext();
        
        // Log device information
        Log.d(TAG, "Device: " + Build.MANUFACTURER + " " + Build.MODEL);
        Log.d(TAG, "Android Version: " + Build.VERSION.RELEASE + " (API " + Build.VERSION.SDK_INT + ")");
        Log.d(TAG, "ABIs: " + Arrays.toString(Build.SUPPORTED_ABIS));
        
        // Check if device architecture is supported
        boolean hasSupportedAbi = false;
        for (String abi : Build.SUPPORTED_ABIS) {
            if (abi.equals("arm64-v8a") || abi.equals("x86_64")) {
                hasSupportedAbi = true;
                break;
            }
        }
        
        if (hasSupportedAbi) {
            Log.d(TAG, "Device has supported ABI");
        } else {
            Log.w(TAG, "Device ABI may not be fully supported: " + Arrays.toString(Build.SUPPORTED_ABIS));
        }
        
        // Test should pass regardless of architecture
        assertTrue("Device compatibility test should pass", true);
    }

    private void copyAssetToStorage(Context context, String assetPath, String destPath) throws IOException {
        AssetManager assetManager = context.getAssets();
        InputStream in = assetManager.open(assetPath);
        File outFile = new File(destPath);
        
        // Create directory if it doesn't exist
        outFile.getParentFile().mkdirs();
        
        FileOutputStream out = new FileOutputStream(outFile);
        byte[] buffer = new byte[1024];
        int read;
        while ((read = in.read(buffer)) != -1) {
            out.write(buffer, 0, read);
        }
        in.close();
        out.close();
    }
}

File: llama_mobile-android-SDK/src/androidTest/java/com/llamamobile/LlamaMobileInstrumentedTests.kt
package com.llamamobile

import android.content.Context
import android.content.res.AssetManager
import androidx.test.platform.app.InstrumentationRegistry
import androidx.test.ext.junit.runners.AndroidJUnit4
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.Assert.*
import java.io.*

/**
 * Instrumented tests for LlamaMobile Android SDK
 * These tests run on actual Android devices or emulators
 */
@RunWith(AndroidJUnit4::class)
class LlamaMobileInstrumentedTests {
    
    private lateinit var context: Context
    private lateinit var assetManager: AssetManager
    
    companion object {
        const val TEST_ASSET_DIR = "grammars"
        const val TEST_GRAMMAR_FILE = "json.gbnf"
        const val TEST_OUTPUT_DIR = "/sdcard/llama_mobile_test"
        const val TEST_MODEL_FILE = "$TEST_OUTPUT_DIR/test_model.gguf"
    }
    
    @Before
    fun setUp() {
        // Get instrumentation context
        context = InstrumentationRegistry.getInstrumentation().context
        assetManager = context.assets
        
        // Create test directory
        val testDir = File(TEST_OUTPUT_DIR)
        if (!testDir.exists()) {
            testDir.mkdirs()
        }
    }
    
    @After
    fun tearDown() {
        // Clean up test files
        val testDir = File(TEST_OUTPUT_DIR)
        if (testDir.exists()) {
            testDir.deleteRecursively()
        }
    }
    
    @Test
    fun testAssetLoading() {
        // Test that we can access the grammar files from assets
        try {
            // List all grammar files
            val grammarFiles = assetManager.list(TEST_ASSET_DIR)
            assertNotNull(grammarFiles)
            assertTrue(grammarFiles!!.isNotEmpty())
            
            // Check if specific grammar file exists
            assertTrue(grammarFiles.contains(TEST_GRAMMAR_FILE))
            
            // Read a grammar file to verify content
            val inputStream = assetManager.open("$TEST_ASSET_DIR/$TEST_GRAMMAR_FILE")
            val reader = BufferedReader(InputStreamReader(inputStream))
            val content = reader.readText()
            reader.close()
            
            assertFalse(content.isEmpty())
            assertTrue(content.contains("json"))
            
        } catch (e: IOException) {
            fail("Asset loading failed: ${e.message}")
        }
    }
    
    @Test
    fun testFileSystemOperations() {
        // Test file system operations that LlamaMobile needs
        try {
            // Create a test file
            val testFile = File(TEST_OUTPUT_DIR, "test.txt")
            testFile.writeText("Hello, LlamaMobile!")
            
            // Verify file exists and contains content
            assertTrue(testFile.exists())
            assertEquals("Hello, LlamaMobile!", testFile.readText())
            
            // Test file copy operation
            val copyFile = File(TEST_OUTPUT_DIR, "test_copy.txt")
            testFile.copyTo(copyFile, overwrite = true)
            assertTrue(copyFile.exists())
            assertEquals(testFile.readText(), copyFile.readText())
            
            // Test file deletion
            copyFile.delete()
            assertFalse(copyFile.exists())
            
        } catch (e: IOException) {
            fail("File system operations failed: ${e.message}")
        }
    }
    
    @Test
    fun testGrammarFileCopyFromAssets() {
        // Test that we can copy grammar files from assets to internal storage
        // This simulates what users would do for structured output
        
        try {
            // Create target directory
            val targetDir = File("${context.filesDir}/grammars")
            if (!targetDir.exists()) {
                targetDir.mkdirs()
            }
            
            // Copy grammar file from assets
            val targetFile = File(targetDir, TEST_GRAMMAR_FILE)
            
            assetManager.open("$TEST_ASSET_DIR/$TEST_GRAMMAR_FILE").use { inputStream ->
                FileOutputStream(targetFile).use { outputStream ->
                    inputStream.copyTo(outputStream)
                }
            }
            
            // Verify file was copied successfully
            assertTrue(targetFile.exists())
            assertTrue(targetFile.length() > 0)
            
        } catch (e: IOException) {
            fail("Grammar file copy failed: ${e.message}")
        }
    }
    
    @Test
    fun testNativeLibraryLoading() {
        // Test that native libraries can be loaded
        // The init block in LlamaMobile should load these automatically
        try {
            // This will fail with UnsatisfiedLinkError if libraries can't be loaded
            System.loadLibrary("llama_mobile")
            System.loadLibrary("llama_mobile_jni")
            
            // If we get here, libraries loaded successfully
            assertTrue(true)
            
        } catch (e: UnsatisfiedLinkError) {
            // Libraries might not be available in test environment, so don't fail
            // Just log the issue
            println("Native libraries not found: ${e.message}")
            // This test is informational only, not critical for CI
        }
    }
    
    @Test
    fun testContextAvailability() {
        // Test that context and environment are properly set up
        assertNotNull(context)
        assertNotNull(assetManager)
        
        // Test that we can access files dir
        assertNotNull(context.filesDir)
        assertTrue(context.filesDir.exists())
        
        // Test that we can access cache dir
        assertNotNull(context.cacheDir)
        assertTrue(context.cacheDir.exists())
    }
    
    @Test
    fun testParameterValidations() {
        // Test that parameter validations work correctly
        
        // Test InitParams validation
        try {
            // This should succeed
            val validParams = LlamaMobile.InitParams(
                modelPath = TEST_MODEL_FILE,
                nCtx = 1024,
                nThreads = 4
            )
            assertNotNull(validParams)
            
        } catch (e: Exception) {
            fail("Valid InitParams threw exception: ${e.message}")
        }
        
        // Test CompletionParams validation
        try {
            // This should succeed
            val validParams = LlamaMobile.CompletionParams(
                prompt = "Hello, world!",
                maxTokens = 10,
                temperature = 0.7f
            )
            assertNotNull(validParams)
            
        } catch (e: Exception) {
            fail("Valid CompletionParams threw exception: ${e.message}")
        }
    }
    
    @Test
    fun testFileAccessPatterns() {
        // Test common file access patterns used by LlamaMobile
        
        // Test 1: Check if we can create a large file (simulating model file)
        val largeTestFile = File(TEST_OUTPUT_DIR, "large_test_file.bin")
        try {
            // Create a 1MB test file
            val buffer = ByteArray(1024 * 1024) // 1MB
            FileOutputStream(largeTestFile).use { fos ->
                fos.write(buffer)
            }
            
            assertTrue(largeTestFile.exists())
            assertEquals(1024 * 1024, largeTestFile.length())
            
        } catch (e: IOException) {
            println("Large file creation failed (may be due to permissions): ${e.message}")
            // Don't fail - this might be due to permissions on some devices
        }
        
        // Test 2: Test directory structure creation
        val nestedDir = File(TEST_OUTPUT_DIR, "nested/dir/structure")
        assertTrue(nestedDir.mkdirs())
        assertTrue(nestedDir.exists())
    }
    
    @Test
    fun testDownloadParamsCompatibility() {
        // Test that download parameters work correctly on Android
        val testUrl = "https://huggingface.co/model"
        val testLocalPath = "$TEST_OUTPUT_DIR/downloaded_model.gguf"
        
        val params = LlamaMobile.DownloadParams(
            url = testUrl,
            localPath = testLocalPath
        )
        
        assertEquals(testUrl, params.url)
        assertEquals(testLocalPath, params.localPath)
        
        // Test with headers (common for authentication)
        val headers = mapOf(
            "User-Agent" to "LlamaMobile-Android/1.0",
            "Authorization" to "Bearer test_token"
        )
        
        val paramsWithHeaders = LlamaMobile.DownloadParams(
            url = testUrl,
            localPath = testLocalPath,
            headers = headers
        )
        
        assertEquals(headers, paramsWithHeaders.headers)
    }
    
    @Test
    fun testDirectoryStructure() {
        // Verify that the SDK directory structure is correct
        
        // Check that the SDK has the expected structure
        val sdkRoot = File(context.packageResourcePath)
        assertNotNull(sdkRoot)
        
        // Check that we can access the APK contents
        assertTrue(sdkRoot.exists())
        assertTrue(sdkRoot.isDirectory)
    }
    
    @Test
    fun testPathOperations() {
        // Test common path operations used by LlamaMobile
        
        // Test that we can convert asset paths to file paths
        val assetPath = "$TEST_ASSET_DIR/$TEST_GRAMMAR_FILE"
        val internalPath = "${context.filesDir}/$assetPath"
        
        val internalFile = File(internalPath)
        val parentDir = internalFile.parentFile
        assertNotNull(parentDir)
        
        // Test path construction
        val modelDir = File(TEST_OUTPUT_DIR, "models")
        val modelFile = File(modelDir, "model.gguf")
        
        assertEquals("model.gguf", modelFile.name)
        assertEquals("models", modelFile.parentFile?.name)
        assertEquals("$TEST_OUTPUT_DIR/models/model.gguf", modelFile.absolutePath)
    }
}
